Fix failure for scalar subquery in value-side of quantified comparison

martint opened the pull request.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/planner/SubqueryPlanner.java:
```diff
@@ -117,7 +117,7 @@ public PlanBuilder handleSubqueries(PlanBuilder builder, Expression expression,
             builder = planInPredicate(builder, cluster, subqueries);
         }
-        for (Cluster<SubqueryExpression> cluster : cluster(builder.getScope(), selectSubqueries(builder, expression, subqueries.getScalarSubqueries()))) {
+        for (Cluster<SubqueryExpression> cluster : cluster(builder.getScope(), selectSubqueries(builder, expression, subqueries.getSubqueries()))) {
```
The rename in this and other classes could be extracted to another commit.
In fact, it logically belongs to this PR: https://github.com/trinodb/trino/pull/7774

martint replied:
Pulled it out into a separate commit.

kasiafi commented on core/trino-main/src/test/java/io/trino/sql/query/TestSubqueries.java:
```diff
@@ -1301,4 +1301,12 @@ public void testCorrelatedUnnestInLateralSubquery()
     }
+
+    @Test
+    public void testQuantifiedComparison()
```
Is there a similar test with in-predicate? If no, could you add one?

martint replied:
Added

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -1903,11 +1900,13 @@ protected Type visitExists(ExistsPredicate node, StackableAstVisitorContext<Cont
 
             Expression subquery = node.getSubquery();
-            process(subquery, context);
+            setExpressionType(subquery, analyzeSubquery((SubqueryExpression) subquery, context));
```
`analyzeSubquery` is always accompanied by `setExpressionType`. Could `analyzeSubquery` do 
```
return(setExpressionType(...))
```
instead?

martint replied:
Yes, that's leftover from a previous experiment.

kasiafi commented on core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -1810,16 +1810,17 @@ protected Type visitInPredicate(InPredicate node, StackableAstVisitorContext<Con
+                subqueryInPredicates.add(NodeRef.of(node));
+                setExpressionType(valueList, analyzeSubquery((SubqueryExpression) valueList, context));
                 coerceToSingleType(context, node, "value and result of subquery must be of the same type for IN expression: %s vs %s", value, valueList);
             }
```
else throw?

martint replied:
Yes, but that's unrelated to this PR. I'll add a separate commit.

kasiafi reviewed (commented):
It looks super clean now. Thanks.

martint committed:
Fix failure for scalar subquery in value-side of quantified comparison

martint committed:
Rename scalarSubqueries to subqueries

Technically, the structure contains the analysis of any SubqueryExpression.
This is also the base for future support of row subqueries.

martint committed:
Handle fall-through case for unexpected class

kasiafi commented:
Looks good to me.
